Adding Python type hints and intergrating mypy for type checking

ris-tlp opened the pull request:
Hopefully this works, in reference to #288 (syncing new python_type_hints branch) and #276 (initial discussion)

Edit:
There we go! I'll make all the commits to this PR now.

cosinekitty commented:
Yes, this looks great. I think we are in good shape now. Just let me know if I can help with any more issues. Thank you so much!

cosinekitty commented on demo/python/astronomy.py:
```diff
@@ -551,18 +551,18 @@ def AngleBetween(a, b):
     if r < 1.0e-8:
-        return BadVectorError()
-    dot = (a.x*b.x + a.y*b.y + a.z*b.z) / r
+        raise BadVectorError()
```
Wow, that was a serious bug! I'm surprised pylint didn't complain about that.

ris-tlp replied:
> Wow, that was a serious bug! I'm surprised pylint didn't complain about that.
Yup, I didn't notice it at first either but mypy immediately flagged it, working out already haha!

ris-tlp commented on generate/template/astronomy.py:
```diff
@@ -2350,7 +2350,7 @@ def _QuadInterp(tm, dt, fa, fm, fb):
     return (t, df_dt)
 
-def Search(func, context, t1, t2, dt_tolerance_seconds):
+def Search(func: Callable[[object, Time], float], context: object, t1: Time, t2: Time, dt_tolerance_seconds: float) -> Optional[Time]:
```
Hey Don, wanted to confirm if this is fine. The documentation here states that T1 and T2 are floats but the tests show that they're both `Time`.

cosinekitty replied:
Yes, you are right. Another good catch! The documentation is wrong. It should say t1 and t2 are Time objects. Feel free to correct the documentation as part of your PR.

ris-tlp commented:
Hey @cosinekitty! I've added type hints to every method/definition that had documentation indicating what type they were. I couldn't interpret the types for the helper methods because I wasn't sure of what they did exactly, but as long as they're being called internally from the documented methods, it should be fine.

cosinekitty commented on generate/pydown/pydown.py:
```diff
@@ -365,7 +365,13 @@ def ConstantsMd(inPythonFileName):
             tokens = line.split()
+            
             if len(tokens) > 3 and tokens[0] == 'from' and tokens[2] == 'import':
+                # from typing import Optional, Union are currently split as
```
Oh, sorry about that. I should fix this part because I can think of cases where it still won't work. But I can do that later.

cosinekitty commented on generate/template/astronomy.py:
```diff
@@ -374,7 +374,7 @@ def _UserDefinedStar(body):
     return None
 
-def DefineStar(body, ra, dec, distanceLightYears):
+def DefineStar(body: Body, ra: float, dec: float, distanceLightYears: float):
```
Is there any advantage to marking the return type with `-> None` here, like you did with some other functions that don't return anything? I don't know enough about Python type hints.

ris-tlp replied:
Yes, if you don't explicitly define the return type as `None`, [it defaults to `Any`](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-any-type) which is pretty much the same as not type hinting at all.

For constructors, [PEP 484](https://peps.python.org/pep-0484/) recommends hinting `None`.

cosinekitty commented on generate/template/astronomy.py:
```diff
@@ -4156,7 +4160,7 @@ def __init__(self, quarter, time):
         return 'MoonQuarter({}, {})'.format(self.quarter, repr(self.time))
 
-def SearchMoonQuarter(startTime):
+def SearchMoonQuarter(startTime: Time):
```
This one returns type MoonQuarter.

cosinekitty commented on generate/template/astronomy.py:
```diff
@@ -4184,7 +4188,7 @@ def SearchMoonQuarter(startTime):
     return MoonQuarter(quarter, time)
 
-def NextMoonQuarter(mq):
+def NextMoonQuarter(mq: MoonQuarter):
```
This one returns MoonQuarter also.

cosinekitty commented on generate/template/astronomy.py:
```diff
@@ -5508,7 +5512,7 @@ def _ToggleAzimuthDirection(az):
 
 
-def VectorFromHorizon(sphere, time, refraction):
+def VectorFromHorizon(sphere: Spherical, time: Time, refraction: Refraction):
```
This one returns Vector. Do you omit the type hint for the return value for the ones where it is obvious?

ris-tlp replied:
Ah no, sorry I missed this one as well.

cosinekitty commented:
Hi @ris-tlp! Thank you so much for all this great work!  I added a few comments. I will take another look at pydown because I realize what I did won't always work. Also, I can help with some of the internal methods. Which ones specifically are confusing? To me, that indicates they really do need some type hints to help better understand them (and maybe some comments too).

ris-tlp commented:
> Hi @ris-tlp! Thank you so much for all this great work! I added a few comments. I will take another look at pydown because I realize what I did won't always work. Also, I can help with some of the internal methods. Which ones specifically are confusing? To me, that indicates they really do need some type hints to help better understand them (and maybe some comments too).

I've resolved all the reviews. Sorry about a few of the missing return types, must have missed them but I've added them now. The internal methods are largely the ones that are doing a bunch of calculations. For example, methods like `_InternalSearchAltitude` have a lot of parameters, some types I can infer but many are difficult to without context.

cosinekitty commented:
This looks good. Thank you so much! Later today I will merge the PR and fix the problems in pydown.

cosinekitty commented:
FYI, I ran into some kind of version difference that caused type hints to be expressed differently. This caused my unit tests to fail:

https://github.com/cosinekitty/astronomy/actions/runs/4218400081/jobs/7322884160

The specific problem is that the generated documentation is a little different in GitHub Actions than it is on my local system. After running the unit tests, I have a sanity check that there are no local changes from git's point of view. Any such changes indicate something is wrong with the reliability and determinism of my build process. Here is an example of what failed in this run:

```
### Search(func: Callable[[object, astronomy.Time], float], context: object, t1: astronomy.Time, t2: astronomy.Time, dt_tolerance_seconds: float) -&gt; Optional[astronomy.Time]
```

becomes:

```
### Search(func: Callable[[object, astronomy.Time], float], context: object, t1: astronomy.Time, t2: astronomy.Time, dt_tolerance_seconds: float) -&gt; Union[astronomy.Time, NoneType]
```

It looks like `Optional[x]` is equivalent to `Union[x, NoneType]`, and different versions of something (probably the module inspection/reflection code in Python) are causing the type hints to be reported differently. I could to try using `Union` instead of `Optional` everywhere to see if that clears things up. Or I could try to modify pydown to recognize this as a special case and always force it to be Optional. Do you have any opinion on this?
